### PR TITLE
8301306: java/net/httpclient/* fail with -Xcomp

### PR DESCRIPTION
**Pull request by sunny868**
Hi all,
I think `BackgroundCompilation` should not be set to false when `-Xcomp` be used, which causes the java thread to block for longer, then causing the following tests failed frequently on the AArch64 and LoongArch64 architecture.

java/net/vthread/HttpALot.java (release -Xcomp)
java/net/httpclient/ByteArrayPublishers.java （fastdebug -Xcomp）
java/net/httpclient/ManyRequestsLegacy.java （fastdebug -Xcomp）
java/net/httpclient/HttpClientLocalAddrTest.java （fastdebug -Xcomp）
java/net/httpclient/SpecialHeadersTest.java  (fastdebug -Xcomp）

This PR fix the issue, Please help review it.

Thanks.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301306](https://bugs.openjdk.org/browse/JDK-8301306): java/net/httpclient/* fail with -Xcomp

### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12282/head:pull/12282` \
`$ git checkout pull/12282`

Update a local copy of the PR: \
`$ git checkout pull/12282` \
`$ git pull https://git.openjdk.org/jdk pull/12282/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12282`

View PR using the GUI difftool: \
`$ git pr show -t 12282`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12282.diff">https://git.openjdk.org/jdk/pull/12282.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/runtime/arguments.cpp [hunk 1]
```diff
@@ -1437,7 +1437,6 @@ void Arguments::set_mode_flags(Mode mode) {
     break;
   case _comp:
     UseInterpreter           = false;
-    BackgroundCompilation    = false;
```
Comment (dcubed-ojdk) [inline src/hotspot/share/runtime/arguments.cpp:1440]: Disabling `BackgroundCompilation` when -Xcomp is specified is not the right
answer here. When `-Xcomp` is specified, `-Xbatch` needs to be turned off
and I believe that's what old L1440 is doing.

**Review by dcubed-ojdk — CHANGES_REQUESTED** [review]
This is not the right fix for this kind of problem.

File: src/hotspot/share/runtime/arguments.cpp (hunk 1 above)
Comment (vnkozlov) [reply src/hotspot/share/runtime/arguments.cpp:1440]: Dan is right, you  can't change the definition of -Xcomp.

But `-Xbatch` is `-XX:-BackgroundCompilation`. So the second Dan's statement is incorrect.

**Comment by AlanBateman** [comment]
There was some discussion about these test failures in the JBS issue and I think @sunny868 has accepted that they need to be investigated to see if they are test reliability issues or something else. As always, running tests with -Xcomp can often challenge timeouts and assumption of progress in tests.

**Comment by vnkozlov** [comment]
> There was some discussion about these test failures in the JBS issue and I think @sunny868 has accepted that they need to be investigated to see if they are test reliability issues or something else. As always, running tests with -Xcomp can often challenge timeouts and assumption of progress in tests.

Yes, -Xcomp can cause long time execution since it we compile all executed methods. But it should not affect execution unless Java code depends on time itself which it should not do.

File: test/jdk/java/net/vthread/HttpALot.java [hunk 2]
```diff
@@ -35,7 +35,7 @@
  * @run main/othervm/timeout=600
  *     -Dsun.net.httpserver.nodelay=true
  *     -Dsun.net.client.defaultConnectTimeout=5000
- *     -Dsun.net.client.defaultReadTimeout=5000
+ *     -Dsun.net.client.defaultReadTimeout=50000
```
Comment (AlanBateman) [inline test/jdk/java/net/vthread/HttpALot.java:38]: For this one, it might be simpler to just drop the setting of the defaultReadTimeout.

**Comment by sunny868** [comment]
The following two tests have the same issue, can I submit the fix code to the same patch?
java/rmi/server/UnicastRemoteObject/serialFilter/FilterUROTest.java and jdk/incubator/concurrent/StructuredTaskScope/StructuredTaskScopeTest.java

**Comment by AlanBateman** [comment]
> The following two tests have the same issue, can I submit the fix code to the same patch?
> java/rmi/server/UnicastRemoteObject/serialFilter/FilterUROTest.java and jdk/incubator/concurrent/StructuredTaskScope/StructuredTaskScopeTest.java

StructuredTaskScopeTest runs 46 tests, and runs them for both platform and virtual threads. Which test is failing? There is no information in the JBS issue on this.

**Review by dfuch — COMMENTED** [review]
The proposed changes look reasonable to me.

**Comment by sunny868** [comment]
> 



> > The following two tests have the same issue, can I submit the fix code to the same patch?
> > java/rmi/server/UnicastRemoteObject/serialFilter/FilterUROTest.java and jdk/incubator/concurrent/StructuredTaskScope/StructuredTaskScopeTest.java
> 
> StructuredTaskScopeTest runs 46 tests, and runs them for both platform and virtual threads. Which test is failing? There is no information in the JBS issue on this.
test StructuredTaskScopeTest.testInterruptJoinUntil1(java.lang.ThreadBuilders$PlatformThreadFactory@52034683): failure

**Comment by sunny868** [comment]
> > The following two tests have the same issue, can I submit the fix code to the same patch?
> > java/rmi/server/UnicastRemoteObject/serialFilter/FilterUROTest.java and jdk/incubator/concurrent/StructuredTaskScope/StructuredTaskScopeTest.java
> 
> StructuredTaskScopeTest runs 46 tests, and runs them for both platform and virtual threads. Which test is failing? There is no information in the JBS issue on this.

test StructuredTaskScopeTest.testInterruptJoinUntil1(java.lang.ThreadBuilders$PlatformThreadFactory@6e8e0c75): failure
I'll set up a in the JBS issue to fix these two test cases later.

File: test/jdk/java/net/vthread/HttpALot.java (hunk 2 above)
Comment (sunny868) [reply test/jdk/java/net/vthread/HttpALot.java:38]: Done

**Comment by sunny868** [comment]
RFR, thanks

**Review by AlanBateman — APPROVED** [review]
Thanks for the update and creating JBS issues/PRs for the -Xcomp vs. timeout issues in tests in other areas.

**Comment by sunny868** [comment]
@AlanBateman Could you please sponsor it for me?

**Review by dfuch — APPROVED** [review]
LGTM

File: test/jdk/java/net/httpclient/SpecialHeadersTest.java [hunk 3]
```diff
@@ -30,6 +30,7 @@
  * @build jdk.httpclient.test.lib.common.HttpServerAdapters
  *        jdk.httpclient.test.lib.http2.Http2TestServer
  *        jdk.test.lib.net.SimpleSSLContext
+ * @requests (vm.compMode != "Xcomp")
```
Comment (dfuch) [inline test/jdk/java/net/httpclient/SpecialHeadersTest.java:33]: I didn't notice this change while reviewing - it is causing trouble in the CI.

Comment (RogerRiggs) [reply test/jdk/java/net/httpclient/SpecialHeadersTest.java:33]: Is PR #12399 related?

Comment (dfuch) [reply test/jdk/java/net/httpclient/SpecialHeadersTest.java:33]: yes - is that RMI test in https://github.com/openjdk/jdk/pull/12399 using an HttpServer?

Comment (dfuch) [reply test/jdk/java/net/httpclient/SpecialHeadersTest.java:33]: It seems that here `@requests` should have been `@requires` ; proposed fix https://github.com/openjdk/jdk/pull/12409

Comment (dfuch) [reply test/jdk/java/net/httpclient/SpecialHeadersTest.java:33]: I doubt that what is proposed in https://github.com/openjdk/jdk/pull/12399 is the right fix for the issue then. I commented there. Still looking for a reviewer for https://github.com/openjdk/jdk/pull/12409 BTW ;-)

Comment (sunny868) [reply test/jdk/java/net/httpclient/SpecialHeadersTest.java:33]: I'm sorry for my low-level typo ;-(

**Pull request closed without merging**